### PR TITLE
Migrations

### PR DESCRIPTION
**Pull request by h-vetinari**
Catch up on all outstanding migrations.

Closes #9 
Closes #34 
Closes #35 
Closes #44

Also build for each supported arrow version, rather than just one. That way, hdk stays compatible with the rest of the ecosystems and up-to-date with the newest versions (modulo the migration PRs of course). We could reduce the explosion of the CI matrix a bit if we decided to only support the 2 newest arrow versions here, for example...

### CODE REVIEW

**Comment by h-vetinari** [comment]
Ping @conda-forge/hdk 
Any thoughts on this PR? Or more importantly: any objections?

**Comment by h-vetinari** [comment]
@conda-forge/hdk
Last call for comments, otherwise this goes in at the beginning of next week.

**Comment by leshikus** [comment]
I believe we better can keep Ilya in the contributor list, he is quite active in HDK project, https://github.com/ienkovich

**Comment by Garra1980** [comment]
> @conda-forge/hdk Last call for comments, otherwise this goes in at the beginning of next week.

Sorry for delay, we want to issue the release first (today/tomorrow) and then review this PR. Thanks!

**Comment by h-vetinari** [comment]
> I believe we better can keep Ilya in the contributor list, he is quite active in HDK project, https://github.com/ienkovich

As I wrote in https://github.com/conda-forge/hdk-feedstock/pull/37/commits/56f4de3809314f59424ccf9e7940073951cd15fc:
> They never accepted any of invitations to join the organisation,
so instead of the bot resending an invite on every merge, remove
them. They're most welcome to rejoin at any point.

He's most welcome to be a maintainer, but then accept the invitation please.

> Sorry for delay, we want to issue the release first (today/tomorrow) and then review this PR. 

Ideally a new release (from the POV of the feedstock) comes on top of this PR, but now that I've heard back, I'm happy to wait a bit (and rebase this PR later if unavoidable).

**Comment by leshikus** [comment]
Some more thoughts regarding this PR. It seems testing takes considerably more time. Thus it may make sense to limit a number of configurations to most popular ones.

**Comment by alexbaden** [comment]
Hi @h-vetinari - 

Thanks for your patience while we got our release out. It looks like this PR needs a rebase. We do have some concerns about the CI explosion, but I agree with your point that having multiple compatible versions tested makes for an easier integration with other packages in the community. That being said, we dropped support for Arrow 10 and our minimum supported Arrow version is 11, so if you could make that change we would appreciate it. That will also reduce the CI explosion somewhat.

**Comment by h-vetinari** [comment]
> We do have some concerns about the CI explosion, but I agree with your point that having multiple compatible versions tested makes for an easier integration with other packages in the community

It's a fair point. I dropped the CUDA 11.8 migration, as that's only really necessary for expanded architecture support. Otherwise, the CUDA 11.2 builds should be compatible with all other 11.x drivers.

> That being said, we dropped support for Arrow 10 and our minimum supported Arrow version is 11, so if you could make that change we would appreciate it.

Great 👍

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:106]: Sidenote: out of curiosity, do you really need to depend on the newest java version? This is going to make hdk impossible to co-install with a relatively large set of java-related packages. Not critical, but curious (on the openjdk feedstock we mainly focus on the LTS versions, but there's been some interest lately also in the intermediate ones).

**Comment by leshikus** [comment]
the code looks good and clean for me, I especially like .bat file changes

File: recipe/meta.yaml
Comment (Garra1980) [inline recipe/meta.yaml:86]: How do all think about using >=3.40.0 here? Looks like this is the only place when we have such a strict version requirement

Comment (leshikus) [reply recipe/meta.yaml:86]: this probably comes from the previous version, >= should work as well IMO

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:106]: Well, who's keeping them from depending on the latest java version on their side? That shouldn't be reflected in the recipe. Besides, java 11 & 17 still see regular releases, and will receive the same amount of security fixes.

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:86]: OK. libsqlite is actually [part](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/800edfcc759a8f7ac1055290bb72191fa08acc7e/recipe/conda_build_config.yaml#L525-L526) of the global pinning and has a very compatible ABI-policy. So generally relying on that will always pull in the newest libsqlite upon build time (say 3.x) and then generate a dependency on `libsqlite >=3.x,<4` at runtime through a [run-export](https://github.com/conda-forge/sqlite-feedstock/blob/ff23d3eda65af793564a1794250f76864cacc985/recipe/meta.yaml#L39).

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:108]: Let's remove this. If hdk just needs _some_ openjdk, then any consumer can add a constraint on `openjdk 20` or whatever they want.
```suggestion
        - openjdk
```
or, if there's a lower bound:

```suggestion
        - openjdk >=11
```

**Comment by h-vetinari** [comment]
@conda-forge/hdk, I'd like to merge this soon. Please let me know what lower bound of openjdk is actually necessary for correct functioning of the package, and if you explicitly want to carry a lower-bound for libsqlite (this would only be for extra clarity; the build already pulls in the newest version by default anyway).

File: recipe/meta.yaml
Comment (Garra1980) [reply recipe/meta.yaml:108]: cc @vlad-penkin

Comment (h-vetinari) [reply recipe/meta.yaml]: I will merge this on the weekend unless there are any further comments until then.

**Review by alexbaden — APPROVED** [review]
Java version pinning looks fine for now, if needed we can tweak it later. Otherwise I think this is good to go. Thanks for all your work on this and my apologies for the initial delays on our side.

File: recipe/meta.yaml
Comment (vlad-penkin) [reply recipe/meta.yaml]: @h-vetinari few comments on openjdk pinning

- pyhdk targets python ecosystem, no interop with java ecosystem. openjdk is only needed to run calcite internally.
- java ecosystem use cases in conda-forge are quite limited and primarily around the bazel, which is needed for builds.
- older openjdk versions in conda-forge tends to have more vulnerabilities, even though 11 and 17 are LTS it's also about the cadence of the updates of the upstream binary builds and subsequent feedstock with repack updates
- we also want to ensure that pyhdk works with all openjdk versions in range, previously we had issues with the jni in older versions.

Comment (h-vetinari) [reply recipe/meta.yaml]: > pyhdk targets python ecosystem, no interop with java ecosystem.

That's fair, though you don't know what packages users will combine in their environments; if anyone needs another package that has a dependence on an LTS openjdk, this pin would prevent them from installing that package side-by-side with hdk, which is not great.

> older openjdk versions in conda-forge tends to have more vulnerabilities, even though 11 and 17 are LTS it's also about the cadence of the updates of the upstream binary builds and subsequent feedstock with repack updates

I honestly doubt that LTS versions have more vulnerabilities, or for longer. Upstream releases builds pretty much concurrently for all supported versions every ~2 months, and the rebuilders (like temurin) that we rely on publish their versions pretty immediately afterwards. I'm also keeping an eye on this on the openjdk feedstock, and currently everything is on its newest supported version.

If you notice an issue, best would be to open an issue (or better: a PR) to the openjdk feedstock.

> we also want to ensure that pyhdk works with all openjdk versions in range, previously we had issues with the jni in older versions.

Sounds great! 👍

**Pull request merged**
